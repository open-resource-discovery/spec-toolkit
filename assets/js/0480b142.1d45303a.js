"use strict";(self.webpackChunk_open_resource_discovery_spec_toolkit_website=self.webpackChunk_open_resource_discovery_spec_toolkit_website||[]).push([[8070],{7208:(e,t,s)=>{s.r(t),s.d(t,{assets:()=>c,contentTitle:()=>n,default:()=>h,frontMatter:()=>r,metadata:()=>i,toc:()=>d});const i=JSON.parse('{"id":"faq","title":"FAQ","description":"Q: Should I write the spec in YAML or JSON","source":"@site/docs/faq.md","sourceDirName":".","slug":"/faq","permalink":"/spec-toolkit/faq","draft":false,"unlisted":false,"tags":[],"version":"current","sidebarPosition":22,"frontMatter":{"title":"FAQ","sidebar_position":22},"sidebar":"defaultSidebar","previous":{"title":"Best Practices","permalink":"/spec-toolkit/best-practices"}}');var a=s(4848),o=s(8453);const r={title:"FAQ",sidebar_position:22},n="FAQ",c={},d=[{value:"Q: Should I write the spec in YAML or JSON",id:"q-should-i-write-the-spec-in-yaml-or-json",level:4},{value:"Q: Do you have a library for language XYZ?",id:"q-do-you-have-a-library-for-language-xyz",level:4}];function l(e){const t={a:"a",h1:"h1",h4:"h4",header:"header",p:"p",strong:"strong",...(0,o.R)(),...e.components};return(0,a.jsxs)(a.Fragment,{children:[(0,a.jsx)(t.header,{children:(0,a.jsx)(t.h1,{id:"faq",children:"FAQ"})}),"\n",(0,a.jsx)(t.h4,{id:"q-should-i-write-the-spec-in-yaml-or-json",children:"Q: Should I write the spec in YAML or JSON"}),"\n",(0,a.jsxs)(t.p,{children:["Both is possible. YAML has some ",(0,a.jsx)(t.a,{href:"https://sidneyliebrand.medium.com/the-greatnesses-and-gotchas-of-yaml-5e3377ef0c55",children:"quirks that you need to be aware of"}),"."]}),"\n",(0,a.jsxs)(t.p,{children:["The advantage over JSON is that it's easy to add comments and especially to have information ",(0,a.jsx)(t.strong,{children:"reuse"})," through YAML anchors for merging data.\nWhen writing the spec, it's very important to not duplicate information where possible, to avoid inconsistencies."]}),"\n",(0,a.jsx)(t.h4,{id:"q-do-you-have-a-library-for-language-xyz",children:"Q: Do you have a library for language XYZ?"}),"\n",(0,a.jsx)(t.p,{children:"Spec JSON Schema itself is described as standard JSON Schema and therefore the entire JSON Schema ecosystem can be used.\nOf course the Spec-Toolkit specific JSON Schema additions will not be supported by general purpose libraries.\nThe specs written with the spec-toolkit are also exported as regular JSON Schema."}),"\n",(0,a.jsxs)(t.p,{children:["Consider using standard JSON Schema validators and converter tools like ",(0,a.jsx)(t.a,{href:"https://quicktype.io/",children:"quicktype"}),".\nVSCode and other IDEs also provide validation and code intelligence by pointing to a JSON Schema file in the root of a JSON document.\nFor YAML, the ",(0,a.jsx)(t.a,{href:"https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml",children:"Red Hat YAML"})," extension will be necessary."]})]})}function h(e={}){const{wrapper:t}={...(0,o.R)(),...e.components};return t?(0,a.jsx)(t,{...e,children:(0,a.jsx)(l,{...e})}):l(e)}}}]);