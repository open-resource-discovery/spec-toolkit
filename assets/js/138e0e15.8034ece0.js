"use strict";(self.webpackChunk_open_resource_discovery_spec_toolkit_website=self.webpackChunk_open_resource_discovery_spec_toolkit_website||[]).push([[4921],{1597:e=>{e.exports=JSON.parse('{"name":"@easyops-cn/docusaurus-search-local","id":"default"}')}}]);