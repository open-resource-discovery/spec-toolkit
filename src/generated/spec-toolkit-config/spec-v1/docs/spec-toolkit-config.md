---
title: "Spec-Toolkit CLI"
sidebar_position: "5"
description: "Describes the technical interface / schema for the Spec-Toolkit CLI configuration."
---

# Spec-Toolkit CLI

This documentation explains how to use the Spec-Toolkit CLI tool. However, most will be done via the [configuration file](#configuration-file-interface-documentation).

## Prerequisite

1. You have installed Node v22 or higher and npm v10 or higher from: [nodejs](https://nodejs.org/en/download/).

1. You have installed the CLI tool on your local machine.

   ```bash
   npm install -g @open-resource-discovery/spec-toolkit
   ```

1. You have prepared a local [configuration file](#configuration-file-interface-documentation).

## CLI signature

Use the CLI tool via [npx](https://docs.npmjs.com/cli/v11/commands/npx).

```sh
# By default it will look for a `./spec-toolkit.config.json` config file in the CWD
npx @open-resource-discovery/spec-toolkit

# Possible to pass a configuration file by file path
npx @open-resource-discovery/spec-toolkit -c ./spec-toolkit.config.json

```

## CLI options

| Name | Value              | Description                                           |
| ---- | ------------------ | ----------------------------------------------------- |
| `-c` | `<configFilePath>` | Configuration file, containing a valid JSON document. |

## Configuration File Interface Documentation

Spec-Toolkit CLI root level configuration document (`spec-toolkit.config.json`), which contains:

- Global config
- Array of all specs JSON Schema files as input and their configuration
- Optional: Array of all spec extensions and their configuration
- Optional: Add plugin config for additional output formats (like diagrams, example files, typescript types).

## Schema Definitions

* The root schema of the document is [Spec-Toolkit Configuration Document](#spec-toolkit-configuration-document)
* The interface is available as JSON Schema: [spec-toolkit-config.schema.json](https://open-resource-discovery.github.io/spec-toolkit/spec-v1/spec-toolkit-config.schema.json#).


### Spec-Toolkit Configuration Document

**Type**: Object(<a href="#spec-toolkit-configuration-document_$schema">$schema</a>, <a href="#spec-toolkit-configuration-document_$id">$id</a>, <a href="#spec-toolkit-configuration-document_generalconfig">generalConfig</a>, <a href="#spec-toolkit-configuration-document_outputpath">outputPath</a>, <a href="#spec-toolkit-configuration-document_docsconfig">docsConfig</a>, <a href="#spec-toolkit-configuration-document_plugins">plugins</a>)

| Property | Type | Description |
| -------- | ---- | ----------- |
|<div className="interface-property-name anchor" id="spec-toolkit-configuration-document_$schema">$schema<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#spec-toolkit-configuration-document_$schema" title="#spec-toolkit-configuration-document.$schema"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">Link to JSON Schema for this spec-toolkit configuration document.<br/>Adding this helps with automatic validation and code intelligence in some editors / IDEs.<br/><br/>See https://tour.json-schema.org/content/06-Combining-Subschemas/02-id-and-schema<br/><hr/>**JSON Schema Format**: `uri-reference`<br/>**Recommended Values**: <ul><li>`"https://github.com/open-resource-discovery/spec-toolkit/spec-v1/spec-toolkit.schema.json#"`</li></ul></div>|
|<div className="interface-property-name anchor" id="spec-toolkit-configuration-document_$id">$id<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#spec-toolkit-configuration-document_$id" title="#spec-toolkit-configuration-document.$id"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">Optional URI for this document, that can acts as an ID or as location to retrieve the document.<br/><br/>See https://tour.json-schema.org/content/06-Combining-Subschemas/02-id-and-schema<br/><hr/>**JSON Schema Format**: `uri-reference`</div>|
|<div className="interface-property-name anchor" id="spec-toolkit-configuration-document_generalconfig">generalConfig<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#spec-toolkit-configuration-document_generalconfig" title="#spec-toolkit-configuration-document.generalConfig"></a></div>|<div className="interface-property-type">[GeneralConfig](#generalconfig)</div>|<div className="interface-property-description">General configuration for the spec-toolkit.<br/>This is optional and can be omitted if no general configuration is needed.</div>|
|<div className="interface-property-name anchor" id="spec-toolkit-configuration-document_outputpath">outputPath<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#spec-toolkit-configuration-document_outputpath" title="#spec-toolkit-configuration-document.outputPath"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The output path where the generated files will be written to.<br/><br/>This is relative to the location of the configuration document.</div>|
|<div className="interface-property-name anchor" id="spec-toolkit-configuration-document_docsconfig">docsConfig<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#spec-toolkit-configuration-document_docsconfig" title="#spec-toolkit-configuration-document.docsConfig"></a></div>|<div className="interface-property-type">Array&lt;[SpecConfig](#specconfig) \| [SpecExtensionConfig](#specextensionconfig)&gt;</div>|<div className="interface-property-description">Configuration for the documentation generation.<hr/>**Array Constraint**: MUST have at least 1 items</div>|
|<div className="interface-property-name anchor" id="spec-toolkit-configuration-document_plugins">plugins<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#spec-toolkit-configuration-document_plugins" title="#spec-toolkit-configuration-document.plugins"></a></div>|<div className="interface-property-type">Array&lt;[PluginConfigData](#pluginconfigdata)&gt;</div>|<div className="interface-property-description">Array of optional plugins to be used for the generation.<br/>Each configured plugin will generate additional output files depending on his specific scope.</div>|


### GeneralConfig

This is the general configuration for the spec-toolkit.
This is optional and can be omitted if no general configuration is needed.

**Type**: Object(<a href="#generalconfig_tstypeexportexcludejsfileextension">tsTypeExportExcludeJsFileExtension</a>)

| Property | Type | Description |
| -------- | ---- | ----------- |
|<div className="interface-property-name anchor" id="generalconfig_tstypeexportexcludejsfileextension">tsTypeExportExcludeJsFileExtension<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#generalconfig_tstypeexportexcludejsfileextension" title="GeneralConfig.tsTypeExportExcludeJsFileExtension"></a></div>|<div className="interface-property-type">boolean</div>|<div className="interface-property-description">If set to `true`, the generated TypeScript types will exclude the `.js` file extension in the export statements.<br/>The '.js' file extension is useful for compatibility with Node.js ESM modules but in Node.js commonjs modules it is not needed.<hr/>**Default Value**: `true`</div>|


### DocsConfigItem

Definition of a docsConfig item.

**Type**: 
[SpecConfig](#specconfig) \| [SpecExtensionConfig](#specextensionconfig)<br/>
**Type**: Object(<a href="#docsconfigitem_type">type</a>)

| Property | Type | Description |
| -------- | ---- | ----------- |
|<div className="interface-property-name anchor" id="docsconfigitem_type">type<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#docsconfigitem_type" title="DocsConfigItem.type"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The docsConfig item type.<br/>It's value is been used as a _discriminator_ to distinguish the matching schema that should be further validated.<hr/>**Allowed Values**: <ul><li>`"spec"`</li><li>`"specExtension"`</li></ul></div>|


### SpecConfig

This is the configuration for a JSON Schema specification.

**Type**: Object(<a href="#specconfig_type">type</a>, <a href="#specconfig_id">id</a>, <a href="#specconfig_sourcefilepath">sourceFilePath</a>, <a href="#specconfig_sourceintrofilepath">sourceIntroFilePath</a>, <a href="#specconfig_sourceoutrofilepath">sourceOutroFilePath</a>, <a href="#specconfig_examplesfolderpath">examplesFolderPath</a>, <a href="#specconfig_mdfrontmatter">mdFrontmatter</a>)

| Property | Type | Description |
| -------- | ---- | ----------- |
|<div className="interface-property-name anchor" id="specconfig_type">type<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#specconfig_type" title="SpecConfig.type"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">Type is used to identify the type of the configuration.<br/><hr/>**Constant Value**: `spec`</div>|
|<div className="interface-property-name anchor" id="specconfig_id">id<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#specconfig_id" title="SpecConfig.id"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The ID of the specification.<br/>This is used to identify the specification in the generated documentation.<br/></div>|
|<div className="interface-property-name anchor" id="specconfig_sourcefilepath">sourceFilePath<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#specconfig_sourcefilepath" title="SpecConfig.sourceFilePath"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The path to the source file of the JSON Schema specification.<br/>This is used to generate the documentation for the specification.<br/><br/>It can be a `.yaml` or `.json` file.<br/><br/>The JSON Schema specification MUST NOT contain objects nested in objects.<br/>Example:<br/>  <code>type: object<br/>    properties:<br/>     &nbsp;&nbsp;fullName:<br/>        &nbsp;&nbsp;&nbsp;&nbsp;type: object<br/>        &nbsp;&nbsp;&nbsp;&nbsp;properties:<br/>          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;firstName:<br/>            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;type: string<br/>          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;lastName:<br/>            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;type: string<br/>      &nbsp;&nbsp;age:<br/>        &nbsp;&nbsp;&nbsp;&nbsp;type: integer<br/>  </code><br/>MUST be replaced by:<br/>  <code>type: object<br/>    properties:<br/>     &nbsp;&nbsp;fullName:<br/>        &nbsp;&nbsp;&nbsp;&nbsp;$ref: "#/definitions/FullName"<br/>      &nbsp;&nbsp;age:<br/>        &nbsp;&nbsp;&nbsp;&nbsp;type: integer<br/>  </code><br/></div>|
|<div className="interface-property-name anchor" id="specconfig_sourceintrofilepath">sourceIntroFilePath<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#specconfig_sourceintrofilepath" title="SpecConfig.sourceIntroFilePath"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The path to the source intro file of the specification.<br/>This is used to generate the documentation for the specification and will be appended at the beginning.<br/></div>|
|<div className="interface-property-name anchor" id="specconfig_sourceoutrofilepath">sourceOutroFilePath<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#specconfig_sourceoutrofilepath" title="SpecConfig.sourceOutroFilePath"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The path to the source outro file of the specification.<br/>This is used to generate the documentation for the specification and will be appended at the end.<br/></div>|
|<div className="interface-property-name anchor" id="specconfig_examplesfolderpath">examplesFolderPath<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#specconfig_examplesfolderpath" title="SpecConfig.examplesFolderPath"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The path to the folder containing the examples for the specification.<br/>This is used to generate the documentation for the specification.<br/><br/>SHOULD contain specification compliant instance example files with extension `.json` or `.jsonc`.<br/></div>|
|<div className="interface-property-name anchor" id="specconfig_mdfrontmatter">mdFrontmatter<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#specconfig_mdfrontmatter" title="SpecConfig.mdFrontmatter"></a></div>|<div className="interface-property-type">[MdFrontmatter](#mdfrontmatter)</div>|<div className="interface-property-description">Frontmatter for the generated documentation.<br/>This is used to generate the markdown frontmatter for the documentation page.</div>|


### SpecExtensionConfig

This is the configuration for a JSON Schema extension specification.

**Type**: Object(<a href="#specextensionconfig_type">type</a>, <a href="#specextensionconfig_id">id</a>, <a href="#specextensionconfig_sourcefilepath">sourceFilePath</a>, <a href="#specextensionconfig_sourceintrofilepath">sourceIntroFilePath</a>, <a href="#specextensionconfig_sourceoutrofilepath">sourceOutroFilePath</a>, <a href="#specextensionconfig_targetdocumentid">targetDocumentId</a>, <a href="#specextensionconfig_mdfrontmatter">mdFrontmatter</a>)

| Property | Type | Description |
| -------- | ---- | ----------- |
|<div className="interface-property-name anchor" id="specextensionconfig_type">type<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#specextensionconfig_type" title="SpecExtensionConfig.type"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">Type is used to identify the type of the configuration.<br/><hr/>**Constant Value**: `specExtension`</div>|
|<div className="interface-property-name anchor" id="specextensionconfig_id">id<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#specextensionconfig_id" title="SpecExtensionConfig.id"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The ID of the specification.<br/>This is used to identify the specification in the generated documentation.<br/></div>|
|<div className="interface-property-name anchor" id="specextensionconfig_sourcefilepath">sourceFilePath<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#specextensionconfig_sourcefilepath" title="SpecExtensionConfig.sourceFilePath"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The path to the source file of the specification.<br/>This is used to generate the documentation for the specification.<br/></div>|
|<div className="interface-property-name anchor" id="specextensionconfig_sourceintrofilepath">sourceIntroFilePath<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#specextensionconfig_sourceintrofilepath" title="SpecExtensionConfig.sourceIntroFilePath"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The path to the source intro file of the specification.<br/>This is used to generate the documentation for the specification and will be appended at the beginning.<br/></div>|
|<div className="interface-property-name anchor" id="specextensionconfig_sourceoutrofilepath">sourceOutroFilePath<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#specextensionconfig_sourceoutrofilepath" title="SpecExtensionConfig.sourceOutroFilePath"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The path to the source outro file of the specification.<br/>This is used to generate the documentation for the specification and will be appended at the end.<br/></div>|
|<div className="interface-property-name anchor" id="specextensionconfig_targetdocumentid">targetDocumentId<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#specextensionconfig_targetdocumentid" title="SpecExtensionConfig.targetDocumentId"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The ID of the target document.<br/>This is used to identify the target document in the generated documentation.<br/></div>|
|<div className="interface-property-name anchor" id="specextensionconfig_mdfrontmatter">mdFrontmatter<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#specextensionconfig_mdfrontmatter" title="SpecExtensionConfig.mdFrontmatter"></a></div>|<div className="interface-property-type">[MdFrontmatter](#mdfrontmatter)</div>|<div className="interface-property-description">Frontmatter for the generated documentation.<br/>This is used to generate the markdown frontmatter for the documentation page.</div>|


### MdFrontmatter

Frontmatter for the generated documentation.
This is used to generate the markdown frontmatter for the documentation page.

| Property | Type | Description |
| -------- | ---- | ----------- |
| Additional Properties<br/><i><code className="regex">^[a-zA-Z0-9_]+$</code></i><br/><span className="optional">OPTIONAL</span> | string | The key of the frontmatter.<br/>This is used to generate the frontmatter for the documentation.<br/><br/><i>Additional properties MUST follow key name regexp pattern</i>: <code className="regex">^[a-zA-Z0-9_]+$</code> |


### PluginConfigData

Configuration for a plugin to be used in the generation.

**Type**: Object(<a href="#pluginconfigdata_packagename">packageName</a>, <a href="#pluginconfigdata_options">options</a>)

| Property | Type | Description |
| -------- | ---- | ----------- |
|<div className="interface-property-name anchor" id="pluginconfigdata_packagename">packageName<br/><span className="mandatory">MANDATORY</span><a className="hash-link" href="#pluginconfigdata_packagename" title="PluginConfigData.packageName"></a></div>|<div className="interface-property-type">string</div>|<div className="interface-property-description">The package name or path to the plugin implementation.<br/><br/>Plugins are implemented as Node.js modules and can be specified by their package name or by a relative path to the spec-toolkit CLI tool when the plugins are maintained in the core spec-toolkit repository.</div>|
|<div className="interface-property-name anchor" id="pluginconfigdata_options">options<br/><span className="optional">OPTIONAL</span><a className="hash-link" href="#pluginconfigdata_options" title="PluginConfigData.options"></a></div>|<div className="interface-property-type">[PluginOptions](#pluginoptions)</div>|<div className="interface-property-description">Optional configuration for the plugin.<br/>This is used to pass additional options to the plugin.</div>|


###### Example Values:


```js
{
  "packageName": "./node_modules/@open-resource-discovery/spec-toolkit/dist/plugin/tabular/index.js"
}
```


```js
{
  "packageName": "./node_modules/@open-resource-discovery/spec-toolkit/dist/plugin/mermaidDiagram/index.js"
}
```


```js
{
  "packageName": "./node_modules/@open-resource-discovery/spec-toolkit/dist/plugin/ums/index.js",
  "options": {
    "metadataPath": "/sap/core/ucl/metadata/ord/v1",
    "overrides": [
      "./document.ums.yaml"
    ]
  }
}
```


```js
{
  "packageName": "@yourNamespace/yourNpmSpecToolkitPluginPackageName"
}
```


### PluginOptions

Optional configuration for the plugin.
This is used to pass additional options to the plugin.

**Type**: Object()

| Property | Type | Description |
| -------- | ---- | ----------- |
| <i>*</i> | | <i>Additional, unspecified properties MAY be added to the object</i>. |

# Spec-Toolkit CLI configuration example

```json
{
  "outputPath": "src/generated/spec-v1",
  "docsConfig": [
    {
      "type": "spec",
      "id": "csn-interop-effective",
      "sourceFilePath": "./spec/v1/CSN-Interop-Effective.schema.yaml",
      "sourceIntroFilePath": "./spec/v1/CSN-Interop-Effective.schema.md",
      "examplesFolderPath": "./spec/v1/examples",
      "mdFrontmatter": {
        "title": "Interface Documentation",
        "sidebar_position": "1",
        "description": "Describes the technical interface / schema for CSN Interop Effective.",
        "toc_max_heading_level": "4"
      }
    },
    {
      "type": "specExtension",
      "id": "aggregation",
      "sourceFilePath": "./spec/v1/annotations/aggregation.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/aggregation.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@Aggregation",
        "sidebar_position": "2",
        "description": "@Aggregation annotations."
      }
    },
    {
      "type": "specExtension",
      "id": "analytics-details",
      "sourceFilePath": "./spec/v1/annotations/analytics-details.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/analytics-details.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@AnalyticsDetails",
        "sidebar_position": "3",
        "description": "@AnalyticsDetails for data analytics use cases."
      }
    },
    {
      "type": "specExtension",
      "id": "consumption",
      "sourceFilePath": "./spec/v1/annotations/consumption.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/consumption.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@Consumption",
        "sidebar_position": "4",
        "description": "@Consumption annotations."
      }
    },
    {
      "type": "specExtension",
      "id": "end-user-text",
      "sourceFilePath": "./spec/v1/annotations/end-user-text.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/end-user-text.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@EndUserText",
        "sidebar_position": "5",
        "description": "@EndUserText annotations for end user UIs."
      }
    },
    {
      "type": "specExtension",
      "id": "entity-relationship",
      "sourceFilePath": "./spec/v1/annotations/entity-relationship.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/entity-relationship.md",
      "sourceOutroFilePath": "./spec/v1/annotations/entity-relationship.outro.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@EntityRelationship",
        "sidebar_position": "6",
        "description": "@EntityRelationship annotations for cross boundary Entity-Relationship IDs and associations."
      }
    },
    {
      "type": "specExtension",
      "id": "object-model",
      "sourceFilePath": "./spec/v1/annotations/object-model.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/object-model.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@ObjectModel",
        "sidebar_position": "7",
        "description": "@ObjectModel annotations."
      }
    },
    {
      "type": "specExtension",
      "id": "odm",
      "sourceFilePath": "./spec/v1/annotations/odm.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/odm.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@ODM",
        "sidebar_position": "8",
        "description": "@ODM for One Domain Model (ODM) related annotations."
      }
    },
    {
      "type": "specExtension",
      "id": "personal-data",
      "sourceFilePath": "./spec/v1/annotations/personal-data.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/personal-data.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@PersonalData",
        "sidebar_position": "9",
        "description": "@PersonalData to annotate DPP relevant information."
      }
    },
    {
      "type": "specExtension",
      "id": "semantics",
      "sourceFilePath": "./spec/v1/annotations/semantics.yaml",
      "sourceIntroFilePath": "./spec/v1/annotations/semantics.md",
      "targetDocumentId": "csn-interop-effective",
      "mdFrontmatter": {
        "title": "@Semantics",
        "sidebar_position": "10",
        "description": "@Semantics annotations."
      }
    }
  ]
}
```