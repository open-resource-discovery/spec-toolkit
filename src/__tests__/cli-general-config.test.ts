import fs from "fs-extra";
import path from "path";
import spawnAsync from "@expo/spawn-async";
import yaml from "js-yaml";
import { SpecToolkitConfigurationDocument } from "../generated/spec-toolkit-config/spec-v1/types/spec-toolkit-config.js";

describe("CLI general config tests", () => {
  const cliBin = "node";
  const cliScriptPath = "./dist/cli.js";
  const tmpTestData = path.join(process.cwd(), "src", "__tests__", "tmpTestData");
  const tmpTestOutput = path.join(process.cwd(), "src", "__tests__", "tmpTestOutput");

  beforeAll(() => {
    // Create test directories and write a valid JSON schema for Person
    fs.ensureDirSync(tmpTestOutput);
    fs.ensureDirSync(tmpTestData);
    const schema = {
      $schema: "http://json-schema.org/draft-07/schema#",
      title: "Person",
      type: "object",
      properties: {
        firstName: { type: "string" },
        lastName: { type: "string" },
      },
      required: ["firstName", "lastName"],
    };
    const schemaPath = path.join(tmpTestData, "person.schema.yaml");
    fs.writeFileSync(schemaPath, yaml.dump(schema), "utf8");
  });

  afterAll(() => {
    fs.removeSync(tmpTestOutput);
    fs.removeSync(tmpTestData);
  });

  describe("Test the general config options", () => {
    describe("Test for tsTypeExportExcludeJsFileExtension", () => {
      test("should successfully omit js file extension when tsTypeExportExcludeJsFileExtension is set to true", async () => {
        const config: SpecToolkitConfigurationDocument = {
          $schema: "https://open-resource-discovery.github.io/spec-toolkit/spec-v1/spec-toolkit-config.schema.json#",
          generalConfig: {
            tsTypeExportExcludeJsFileExtension: true, // set to true
          },
          outputPath: "src/__tests__/tmpTestOutput",
          docsConfig: [
            {
              type: "spec",
              id: "my-spec",
              sourceFilePath: "./src/__tests__/tmpTestData/person.schema.yaml",
            },
          ],
        };
        const configFilePath = tmpTestData.concat("/config.json");
        fs.writeFileSync(configFilePath, JSON.stringify(config), "utf8");

        const cliArguments = [cliScriptPath, "-c", "./src/__tests__/tmpTestData/config.json"];

        const resultPromise = spawnAsync(cliBin, cliArguments);

        try {
          const { stdout, stderr } = await resultPromise;

          expect(stdout).toContain("SUCCESS: Documentation successfully generated to src/__tests__/tmpTestOutput");
          // Check that stderr is empty
          expect(stderr).toEqual("");

          // Read output files and check their content
          const indexTypesFileContent = fs.readFileSync("src/__tests__/tmpTestOutput/types/index.ts").toString();

          expect(indexTypesFileContent).toMatchSnapshot();
        } catch (e) {
          expect(e).toEqual("expect this to never happen because above code should not throw an error");
        }
      });
      test("should append js file extension when tsTypeExportExcludeJsFileExtension is set to false", async () => {
        const config: SpecToolkitConfigurationDocument = {
          $schema: "https://open-resource-discovery.github.io/spec-toolkit/spec-v1/spec-toolkit-config.schema.json#",
          generalConfig: {
            tsTypeExportExcludeJsFileExtension: false, // set to false
          },
          outputPath: "src/__tests__/tmpTestOutput",
          docsConfig: [
            {
              type: "spec",
              id: "my-spec",
              sourceFilePath: "./src/__tests__/tmpTestData/person.schema.yaml",
            },
          ],
        };
        const configFilePath = tmpTestData.concat("/config.json");
        fs.writeFileSync(configFilePath, JSON.stringify(config), "utf8");

        const cliArguments = [cliScriptPath, "-c", "./src/__tests__/tmpTestData/config.json"];

        const resultPromise = spawnAsync(cliBin, cliArguments);

        try {
          const { stdout, stderr } = await resultPromise;

          expect(stdout).toContain("SUCCESS: Documentation successfully generated to src/__tests__/tmpTestOutput");
          // Check that stderr is empty
          expect(stderr).toEqual("");

          // Read output files and check their content
          const indexTypesFileContent = fs.readFileSync("src/__tests__/tmpTestOutput/types/index.ts").toString();

          expect(indexTypesFileContent).toMatchSnapshot();
        } catch (e) {
          expect(e).toEqual("expect this to never happen because above code should not throw an error");
        }
      });

      test("should append js file extension when generalConfig is missing at all", async () => {
        const config: SpecToolkitConfigurationDocument = {
          $schema: "https://open-resource-discovery.github.io/spec-toolkit/spec-v1/spec-toolkit-config.schema.json#",
          outputPath: "src/__tests__/tmpTestOutput",
          docsConfig: [
            {
              type: "spec",
              id: "my-spec",
              sourceFilePath: "./src/__tests__/tmpTestData/person.schema.yaml",
            },
          ],
        };
        const configFilePath = tmpTestData.concat("/config.json");
        fs.writeFileSync(configFilePath, JSON.stringify(config), "utf8");

        const cliArguments = [cliScriptPath, "-c", "./src/__tests__/tmpTestData/config.json"];

        const resultPromise = spawnAsync(cliBin, cliArguments);

        try {
          const { stdout, stderr } = await resultPromise;

          expect(stdout).toContain("SUCCESS: Documentation successfully generated to src/__tests__/tmpTestOutput");
          // Check that stderr is empty
          expect(stderr).toEqual("");

          // Read output files and check their content
          const indexTypesFileContent = fs.readFileSync("src/__tests__/tmpTestOutput/types/index.ts").toString();

          expect(indexTypesFileContent).toMatchSnapshot();
        } catch (e) {
          expect(e).toEqual("expect this to never happen because above code should not throw an error");
        }
      });
    });
  });
});
