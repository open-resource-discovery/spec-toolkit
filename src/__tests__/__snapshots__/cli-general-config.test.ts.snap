// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CLI general config tests Test the general config options Test for tsTypeExportExcludeJsFileExtension should append js file extension when generalConfig is missing at all 1`] = `
"export * from "./my-spec.js";
"
`;

exports[`CLI general config tests Test the general config options Test for tsTypeExportExcludeJsFileExtension should append js file extension when tsTypeExportExcludeJsFileExtension is set to false 1`] = `
"export * from "./my-spec.js";
"
`;

exports[`CLI general config tests Test the general config options Test for tsTypeExportExcludeJsFileExtension should successfully omit js file extension when tsTypeExportExcludeJsFileExtension is set to true 1`] = `
"export * from "./my-spec";
"
`;
